### DISCUSSION TITLE
Getting Started As Developer

### DISCUSSION THREAD

**Discussion by idhamari** [iaBIH/arx#5] [ARX Developer Tutorials]
**Build ARX Source-code using Eclipse**

Check this [video tutorial](https://drive.google.com/file/d/11lFZk71YCDAIUUPjwAObhrOfC93V6w-D/view?usp=sharing).
 
1. Preparing your github repository
   * [Fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo) the [ARX repository](https://github.com/arx-deidentifier/arx)
   * [Clone](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) your forked repository.
   * Create [new branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-and-deleting-branches-within-your-repository) for your development  
2. Preparing your IDE
   * Install [eclipse for jave](https://www.eclipse.org/downloads/packages/release/kepler/sr1/eclipse-ide-java-developers)
   * Create a new java project as follows:
       - File> New
          ![Screenshot from 2022-02-08 09-00-47](https://user-images.githubusercontent.com/28802516/152943872-cabefa63-6305-4886-9991-2499d25d78b2.png)
       - Un-check the default location
       ![Screenshot from 2022-02-08 09-01-21](https://user-images.githubusercontent.com/28802516/152943975-1832d759-aff3-46c2-9ac9-133da7c59d91.png)
       - Browse to your ARX cloned repository then click the "Finish" button
![Screenshot from 2022-02-08 09-02-08](https://user-images.githubusercontent.com/28802516/152944018-8da34574-4c4b-4500-8f47-29a15b4fc3c3.png)
   * Remove libraries for other systems e.g. if you are in windows, remove libs for Linux and mac 
     - todo (screenshots)   
3. Building and running ARX
    * Test [first example](https://github.com/arx-deidentifier/arx/blob/master/src/example/org/deidentifier/arx/examples/Example1.java) and [GUI](https://github.com/arx-deidentifier/arx/blob/master/src/gui/org/deidentifier/arx/gui/Main.java)
    * Read the [doc](https://arx.deidentifier.org/development/api/)
4. Start contribution:
   * Add [arx formatter](https://github.com/iaBIH/arx/files/9248389/arx-code-formatter.zip) to eclipse, this helps unifying the code styling and make the code more readable and easy to understand:
     - Download and extract the file inside the zip file in the link above. 
     - In eclipse, select Window > Preferences
     ![Screenshot from 2022-08-03 08-05-54](https://user-images.githubusercontent.com/28802516/182536546-5391ca43-409f-4cb3-a14e-a57507549426.png)
     - Select Java>Code Style > Formatter then click on the Import button
     ![Screenshot from 2022-08-03 08-07-03](https://user-images.githubusercontent.com/28802516/182536563-54566324-4fcb-418a-9708-b8323b39aa4f.png)
     - Open the extracted downloaded XML file
        ![Screenshot from 2022-08-03 08-07-34](https://user-images.githubusercontent.com/28802516/182536569-f4c38e1d-d139-49ab-b7e6-4a6d9a177f86.png)
    - Click on Apply and Close
![Screenshot from 2022-08-03 08-08-11](https://user-images.githubusercontent.com/28802516/182536580-546338c7-e667-4ae7-a0e6-d0e634995a71.png)
 - Before start contribution, please read [this](https://github.com/arx-deidentifier/arx/blob/master/contributing.md) and [this](https://github.com/arx-deidentifier/arx/blob/master/code_of_conduct.md). 

**Debugging**

Debugging should be similar to other java applications, this is a good [tutorial](https://www.vogella.com/tutorials/EclipseDebugging/article.html) about debugging in eclipse.

**Comment by prasser** [comment]
Most libraries will be on the classpath twice. In addition to removing all libraries for the wrong system, also all libraries that are not located in ./lib/ant must be removed.

**Reply by prasser** [reply]
Also, modern Eclipse asks you whether you want to create a module. You must select "No".

**Comment by prasser** [comment]
A description for building with Maven (instead of ant) would also be great.

**Reply by idhamari** [reply]
Thanks for your feedback. I will consider this in the next video tutorial.

**Reply by idhamari** [reply]
> A description for building with Maven (instead of ant) would also be great.

@prasser 
I tried to build ARX using a recent version of maven i.e. 3.8.4  e.g. mvn compile but I got errors related to swt e.g.: 

      [ERROR] /home/ibr/Downloads/arx-master/src/gui/org/deidentifier/arx/gui/view/impl/menu/DialogComboSelection.java:[26,23] package org.eclipse.swt does not exist

I was able to build ARX API without problem  using 

              mvn compile -Dcore=true 

This is the complete [error output](https://github.com/iaBIH/arx/files/8035524/arx-mvn-log.txt)

**Reply by idhamari** [reply]
I solved the problem above by adding the library manually using eclipse interface. I got another error related to the library we updated last time i.e. de.linearbits.preferences.PreferenceBoolean 

Please correct me @prasser if I am wrong, I think the maven file should be updated as well.
